libraries/AP_Generator/AP_Generator_Loweheiser.h

```diff
@@ -0,0 +1,198 @@
+
+    // parameters for the Loweheiser generator:
+    static const struct AP_Param::GroupInfo var_info[];
+
+    const struct AP_Param::GroupInfo *get_var_info() const override {
+        return var_info;
+    }
+
+    // method to reset the amount of energy remaining in a generator.
+    // This typically means someone has refueled the vehicle without
+    // powering it off, and is indicating that the fuel tank is full.
+    bool reset_consumed_energy() override {
+        last_consumed_mah_reset_value = _consumed_mah;
+        return true;
+    }
+
+private:
+
+    // we store the entirety of the most recent packet for the time being:
+    mavlink_loweheiser_gov_efi_t packet;
+    uint32_t last_packet_received_ms;
+
+    mavlink_command_ack_t ack_packet;
+    uint32_t last_ack_packet_ms;
+    // record the timestamp of the most recently processed ack:
+    uint32_t last_ack_packet_processed_ms;
+    // good_ack is true if the last ack we received from the generator
+    // reported success:
+    bool good_ack;
+
+    // state recorded from the first Loweheiser message we see so we
+    // only process from one source:
+    bool seen_good_message;
+    const class GCS_MAVLINK *mavlink_channel;
+    uint8_t sysid;
+    uint8_t compid;
+    uint8_t efi_index;
+
+    // methods and state to record pilot desired runstate and actual runstate:
+    enum class RunState {
+        STOP = 17,
+        IDLE = 18,
+        WARMING_UP = 19,
+        RUN = 20,
+        COOLING_DOWN = 21,
+    };
+    enum class PilotDesiredRunState {
+        STOP = 65,
+        IDLE = 66,
+        RUN = 67,
+    };
+    PilotDesiredRunState pilot_desired_runstate = PilotDesiredRunState::STOP;
+    // last pilot desired runstate we acted on; forced to an invalid
+    // value so that we action at least once.
+    RunState commanded_runstate = RunState::STOP;  // output is based on this
+    void set_pilot_desired_runstate(PilotDesiredRunState newstate);
+    void update_runstate();
+    const char *runstate_string(PilotDesiredRunState runstate);
+
+    // returns true if the generator should be allowed to move into
+    // the "run" (high-RPM) state:
+    bool generator_ok_to_run() const;
+
+    // returns true if the generator should be allowed to move into
+    // the "stop" state.  This is used to enforce a cool-down period
+    // for the device
+    bool generator_ok_to_stop() const;
+
+    // timestamp of last time we told the user we're waiting for the
+    // generator to change temperature:
+    uint32_t last_waiting_temperature_change_ms;
+
+    // estop_reported is true if we do not need to tell the user of an
+    // estop event if it occurs:
+    bool estop_reported;
+
+    // prepare and send commands to generator:
+    void command_generator();
+
+    // threshold temperature (degC) for when the mavlink
+    // overtemperature bit should be set.  Might want a parameter for
+    // this:
+    static constexpr float temp_for_overtemp_warning = 190;
+
```

Comment:
This should be made into a parameter still I think?  Though I don't think it was 100% decided on the value of 205 deg C. Lowheiser was thinking that was too high I believe.